### Introduction

Data wrangling is a core skill that everyone who works with data should be familiar with since so much of the world's data isn't clean. The repurcussions of ignoring data wrangling before analysis can be so bad and sometimes unimaginable. If we analyze, visualize, or model our data before we wrangle it, our consequences could be making mistakes, missing out on cool insights, and wasting time. So best practices say wrangle. Always.

In this project, we gather data from various sources and in variety of forms. When data is sourced from different sources, there are extreme inconsistencies between them. So inorder to analyse this data, cleaning must be considered first.

The dataset that we will be wrangling (and analyzing and visualizing) is the tweet archive of Twitter user @dog_rates, also known as WeRateDogs.

In [88]:
# import libraries

import pandas as pd
import requests
import time
import json
import tweepy
import os

### Gather Data

#### Gather data 1 - from the downloadable link

In [89]:
# load tweets archieve data
tweets_df = pd.read_csv('twitter-archive-enhanced.csv')
tweets_df.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

#### Gather data 2 - fetch data from the url using requests library

In [90]:
# get image preictions data

url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
response = requests.get(url)
with open(os.path.join('./', url.split('/')[-1]), mode='wb') as file:
    file.write(response.content)

In [91]:
# load image predictions data

images_df = pd.read_csv('image-predictions.tsv', sep='\t')
images_df.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

#### Gather data 3 - More info from WeRateDogs using Twitter API

In [92]:
# twitter api tokens

consumer_key = ''
consumer_secret = ''
access_token = ''
access_secret = ''

auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

api = tweepy.API(auth)

In [93]:
# deleted_ids = []
# with open(os.path.join('./', 'tweet_json.txt'), mode='w') as file:
#     for id in tweets_df.tweet_id:
#         start = time.time()
#         try:
#             data = api.get_status(id, tweet_mode='extended', wait_on_rate_limit=True, wait_on_rate_limit_notify=True)
#             file.write(json.dumps(data._json) + '\n')
#         except:
#             deleted_ids.append(id)
#         end = time.time()
#         print(str(data.id) + "------------" + str(end-start))
# len(deleted_ids)

In [94]:
# read tweet_json.txt and create an appropriate dataframe
df_list = []
with open('tweet_json.txt', 'r') as file:
    lines = file.read().splitlines()
    for line_data in lines:
        json_data = json.loads(line_data)
        tweet_id = json_data['id']
        favorite_count = json_data['favorite_count']
        retweet_count = json_data['retweet_count']
        df_list.append({
            'tweet_id': tweet_id,
            'favorite_count': favorite_count,
            'retweet_count': retweet_count
        })
likes_df = pd.DataFrame(df_list)
likes_df.head()

favorite_count  retweet_count            tweet_id
0           38388           8431  892420643555336193
1           32894           6211  892177421306343426
2           24786           4111  891815181378084864
3           41723           8560  891689557279858688
4           39897           9285  891327558926688256

### Assess Data 

#### Visual Assessment 

In [95]:
tweets_df

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
5     891087950875897856                    NaN                  NaN   
6     890971913173991426                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
8     890609185150312448                    NaN                  NaN   
9     890240255349198849                    NaN                  NaN   
10    890006608113172480                    NaN                  NaN   
11    889880896479866881                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
13    889638837579907072                    NaN                  NaN   
14    889531135344209921                    NaN                  NaN   
15    889278841981685760                    NaN                  NaN   
16    888917238123831296                    NaN                  NaN   
17    888804989199671297                    NaN                  NaN   
18    888554962724278272                    NaN                  NaN   
19    888202515573088257                    NaN                  NaN   
20    888078434458587136                    NaN                  NaN   
21    887705289381826560                    NaN                  NaN   
22    887517139158093824                    NaN                  NaN   
23    887473957103951883                    NaN                  NaN   
24    887343217045368832                    NaN                  NaN   
25    887101392804085760                    NaN                  NaN   
26    886983233522544640                    NaN                  NaN   
27    886736880519319552                    NaN                  NaN   
28    886680336477933568                    NaN                  NaN   
29    886366144734445568                    NaN                  NaN   
...                  ...                    ...                  ...   
2326  666411507551481857                    NaN                  NaN   
2327  666407126856765440                    NaN                  NaN   
2328  666396247373291520                    NaN                  NaN   
2329  666373753744588802                    NaN                  NaN   
2330  666362758909284353                    NaN                  NaN   
2331  666353288456101888                    NaN                  NaN   
2332  666345417576210432                    NaN                  NaN   
2333  666337882303524864                    NaN                  NaN   
2334  666293911632134144                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   
2336  666273097616637952                    NaN                  NaN   
2337  666268910803644416                    NaN                  NaN   
2338  666104133288665088                    NaN                  NaN   
2339  666102155909144576                    NaN                  NaN   
2340  666099513787052032                    NaN                  NaN   
2341  666094000022159362                    NaN                  NaN   
2342  666082916733198337                    NaN                  NaN   
2343  666073100786774016                    NaN                  NaN   
2344  666071193221509120                    NaN                  NaN   
2345  666063827256086533                    NaN                  NaN   
2346  666058600524156928                    NaN                  NaN   
2347  666057090499244032                    NaN                  NaN   
2348  666055525042405380                    NaN                  NaN   
2349  666051853826850816                    NaN         

In [96]:
images_df

tweet_id                                            jpg_url  \
0     666020888022790149    https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1     666029285002620928    https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2     666033412701032449    https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3     666044226329800704    https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4     666049248165822465    https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
5     666050758794694657    https://pbs.twimg.com/media/CT5Jof1WUAEuVxN.jpg   
6     666051853826850816    https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg   
7     666055525042405380    https://pbs.twimg.com/media/CT5N9tpXIAAifs1.jpg   
8     666057090499244032    https://pbs.twimg.com/media/CT5PY90WoAAQGLo.jpg   
9     666058600524156928    https://pbs.twimg.com/media/CT5Qw94XAAA_2dP.jpg   
10    666063827256086533    https://pbs.twimg.com/media/CT5Vg_wXIAAXfnj.jpg   
11    666071193221509120    https://pbs.twimg.com/media/CT5cN_3WEAAlOoZ.jpg   
12    666073100786774016    https://pbs.twimg.com/media/CT5d9DZXAAALcwe.jpg   
13    666082916733198337    https://pbs.twimg.com/media/CT5m4VGWEAAtKc8.jpg   
14    666094000022159362    https://pbs.twimg.com/media/CT5w9gUW4AAsBNN.jpg   
15    666099513787052032    https://pbs.twimg.com/media/CT51-JJUEAA6hV8.jpg   
16    666102155909144576    https://pbs.twimg.com/media/CT54YGiWUAEZnoK.jpg   
17    666104133288665088    https://pbs.twimg.com/media/CT56LSZWoAAlJj2.jpg   
18    666268910803644416    https://pbs.twimg.com/media/CT8QCd1WEAADXws.jpg   
19    666273097616637952    https://pbs.twimg.com/media/CT8T1mtUwAA3aqm.jpg   
20    666287406224695296    https://pbs.twimg.com/media/CT8g3BpUEAAuFjg.jpg   
21    666293911632134144    https://pbs.twimg.com/media/CT8mx7KW4AEQu8N.jpg   
22    666337882303524864    https://pbs.twimg.com/media/CT9OwFIWEAMuRje.jpg   
23    666345417576210432    https://pbs.twimg.com/media/CT9Vn7PWoAA_ZCM.jpg   
24    666353288456101888    https://pbs.twimg.com/media/CT9cx0tUEAAhNN_.jpg   
25    666362758909284353    https://pbs.twimg.com/media/CT9lXGsUcAAyUFt.jpg   
26    666373753744588802    https://pbs.twimg.com/media/CT9vZEYWUAAlZ05.jpg   
27    666396247373291520    https://pbs.twimg.com/media/CT-D2ZHWIAA3gK1.jpg   
28    666407126856765440    https://pbs.twimg.com/media/CT-NvwmW4AAugGZ.jpg   
29    666411507551481857    https://pbs.twimg.com/media/CT-RugiWIAELEaq.jpg   
...                  ...                                                ...   
2045  886366144734445568    https://pbs.twimg.com/media/DE0BTnQUwAApKEH.jpg   
2046  886680336477933568    https://pbs.twimg.com/media/DE4fEDzWAAAyHMM.jpg   
2047  886736880519319552    https://pbs.twimg.com/media/DE5Se8FXcAAJFx4.jpg   
2048  886983233522544640    https://pbs.twimg.com/media/DE8yicJW0AAAvBJ.jpg   
2049  887101392804085760    https://pbs.twimg.com/media/DE-eAq6UwAA-jaE.jpg   
2050  887343217045368832  https://pbs.twimg.com/ext_tw_video_thumb/88734...   
2051  887473957103951883    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2052  887517139158093824  https://pbs.twimg.com/ext_tw_video_thumb/88751...   
2053  887705289381826560    https://pbs.twimg.com/media/DFHDQBbXgAEqY7t.jpg   
2054  888078434458587136    https://pbs.twimg.com/media/DFMWn56WsAAkA7B.jpg   
2055  888202515573088257    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2056  888554962724278272    https://pbs.twimg.com/media/DFTH_O-UQAACu20.jpg   
2057  888804989199671297    https://pbs.twimg.com/media/DFWra-3VYAA2piG.jpg   
2058  888917238123831296    https://pbs.twimg.com/media/DFYRgsOUQAARGhO.jpg   
2059  889278841981685760  https://pbs.twimg.com/ext_tw_video_thumb/88927...   
2060  889531135344209921    https://pbs.twimg.com/media/DFg_2PVW0AEHN3p.jpg   
2061  889638837579907072    https://pbs.twimg.com/media/DFihzFfXsAYGDPR.jpg   
2062  889665388333682689    https://pbs.twimg.com/media/DFi579UWsAAatzw.jpg   
2063  889880896479866881    https://pbs.twimg.com/media/DFl99B1WsA

In [97]:
likes_df

favorite_count  retweet_count            tweet_id
0              38388           8431  892420643555336193
1              32894           6211  892177421306343426
2              24786           4111  891815181378084864
3              41723           8560  891689557279858688
4              39897           9285  891327558926688256
5              20022           3081  891087950875897856
6              11722           2045  890971913173991426
7              64797          18697  890729181411237888
8              27538           4222  890609185150312448
9              31599           7321  890240255349198849
10             30349           7263  890006608113172480
11             27499           4927  889880896479866881
12             47635           9962  889665388333682689
13             26889           4494  889638837579907072
14             14941           2214  889531135344209921
15             25041           5365  889278841981685760
16             28799           4457  888917238123831296
17             25336           4279  888804989199671297
18             19673           3540  888554962724278272
19             21540           3465  888078434458587136
20             29860           5333  887705289381826560
21             45793          11555  887517139158093824
22             68400          18019  887473957103951883
23             33337          10300  887343217045368832
24             30246           5907  887101392804085760
25             34809           7705  886983233522544640
26             11932           3259  886736880519319552
27             22199           4421  886680336477933568
28             20981           3172  886366144734445568
29               116              4  886267009285017600
...              ...            ...                 ...
2311             442            324  666411507551481857
2312             108             40  666407126856765440
2313             167             84  666396247373291520
2314             187             92  666373753744588802
2315             774            566  666362758909284353
2316             219             72  666353288456101888
2317             293            135  666345417576210432
2318             197             90  666337882303524864
2319             498            352  666293911632134144
2320             145             64  666287406224695296
2321             174             76  666273097616637952
2322             102             35  666268910803644416
2323           14227           6533  666104133288665088
2324              80             12  666102155909144576
2325             153             67  666099513787052032
2326             163             73  666094000022159362
2327             117             44  666082916733198337
2328             320            163  666073100786774016
2329             146             60  666071193221509120
2330             471            216  666063827256086533
2331             111             57  666058600524156928
2332             295            141  666057090499244032
2333             433            247  666055525042405380
2334            1210            843  666051853826850816
2335             131             59  666050758794694657
2336             109             40  666049248165822465
2337             298            140  666044226329800704
2338             125             44  666033412701032449
2339             129             47  666029285002620928
2340            2555            508  666020888022790149

[2341 rows x 3 columns]

#### Programmatic Assessment

In [98]:
likes_df.shape, images_df.shape, tweets_df.shape

((2341, 3), (2075, 12), (2356, 17))

In [99]:
tweets_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [100]:
images_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [101]:
likes_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2341 entries, 0 to 2340
Data columns (total 3 columns):
favorite_count    2341 non-null int64
retweet_count     2341 non-null int64
tweet_id          2341 non-null int64
dtypes: int64(3)
memory usage: 54.9 KB


In [102]:
# checks for retweets
tweets_df.query('retweeted_status_id in tweet_id')

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
19    888202515573088257                    NaN                  NaN   
36    885311592912609280                    NaN                  NaN   
68    879130579576475649                    NaN                  NaN   
73    878404777348136964                    NaN                  NaN   
74    878316110768087041                    NaN                  NaN   
91    874434818259525634                    NaN                  NaN   
95    873697596434513921                    NaN                  NaN   
97    873337748698140672                    NaN                  NaN   
109   871166179821445120                    NaN                  NaN   
118   869988702071779329                    NaN                  NaN   
124   868639477480148993                    NaN                  NaN   
132   866816280283807744                    NaN                  NaN   
137   866094527597207552                    NaN                  NaN   
146   863471782782697472                    NaN                  NaN   
155   861769973181624320                    NaN                  NaN   
159   860981674716409858                    NaN                  NaN   
165   860177593139703809                    NaN                  NaN   
171   858860390427611136                    NaN                  NaN   
182   856602993587888130                    NaN                  NaN   
194   855245323840757760                    NaN                  NaN   
204   852936405516943360                    NaN                  NaN   
211   851953902622658560                    NaN                  NaN   
222   849668094696017920                    NaN                  NaN   
230   847978865427394560                    NaN                  NaN   
247   845459076796616705                    NaN                  NaN   
250   845098359547420673                    NaN                  NaN   
260   842892208864923648                    NaN                  NaN   
266   841833993020538882                    NaN                  NaN   
272   840761248237133825                    NaN                  NaN   
273   840728873075638272                    NaN                  NaN   
...                  ...                    ...                  ...   
759   778396591732486144                    NaN                  NaN   
764   777953400541634568                    NaN                  NaN   
767   777641927919427584                    NaN                  NaN   
770   776819012571455488                    NaN                  NaN   
773   776249906839351296                    NaN                  NaN   
778   775898661951791106                    NaN                  NaN   
784   775096608509886464                    NaN                  NaN   
794   773336787167145985                    NaN                  NaN   
800   772615324260794368                    NaN                  NaN   
811   771171053431250945                    NaN                  NaN   
818   770743923962707968                    NaN                  NaN   
822   770093767776997377                    NaN                  NaN   
826   769335591808995329                    NaN                  NaN   
829   768909767477751808                    NaN                  NaN   
833   768554158521745409                    NaN                  NaN   
841   766864461642756096                    NaN                  NaN   
847   766078092750233600                    NaN                  NaN   
860   763167063695355904                    NaN                  NaN   
868   761750502866649088                    NaN                  NaN   
872   761371037149827077                    NaN                  NaN   
890   759566828574212096                    NaN                  NaN   
895   759159934323924993                    NaN                  NaN   
908   757729163776290825                    NaN                  NaN   
926   754874841593970688                    NaN         

In [103]:
sum(tweets_df.expanded_urls.isnull())

59

In [104]:
tweets_df.rating_numerator.unique()

array([  13,   12,   14,    5,   17,   11,   10,  420,  666,    6,   15,
        182,  960,    0,   75,    7,   84,    9,   24,    8,    1,   27,
          3,    4,  165, 1776,  204,   50,   99,   80,   45,   60,   44,
        143,  121,   20,   26,    2,  144,   88])

In [105]:
tweets_df.rating_denominator.unique()

array([ 10,   0,  15,  70,   7,  11, 150, 170,  20,  50,  90,  80,  40,
       130, 110,  16, 120,   2])

### Issues List

#### Quality Issues:
- retweets are included.
- Names of the dogs are wrong (a, an, the, this, etc)
- retweeted_status_timestamp and timestamp are not datetime objects and in_reply_to_status_id, in_reply_to_user_id, retweeted_status_id, retweeted_status_user_id should be str object
- null represented as 'None' in columns 'name', 'doggo', 'floofer', 'pupper','puppo' instead of NaN.
- Ratings are not extracted properly from text column.
- tweet_id - 835246439529840640 has a 0 denominator rating.
- missing tweet images(expanded url).
- number of observations(rows) is not consistent - (images_df: 2075 instead of tweets_df: 2356)

#### Tidiness Issues:
- stage (doggo, floofer, pupper, puppo) in 4 different columns
- All data is not present in one dataset

### Cleaning data:

In [106]:
# copy all dataframes to another variable

tweet_clean = tweets_df.copy()
images_clean = images_df.copy()
likes_clean = likes_df.copy()

#### Define
- Drop the retweets data

#### Clean

In [107]:
# query for retweets

retweets = tweet_clean.query('retweeted_status_id in tweet_id')

In [108]:
# drop the retweets columns

tweet_clean.drop(retweets.index, inplace=True)

#### Test

In [109]:
retweets = tweet_clean.query('retweeted_status_id in tweet_id')
retweets

Empty DataFrame
Columns: [tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, source, text, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo]
Index: []

#### Define
 - Replace the inappropriate names of the dogs to None

#### Clean

In [110]:
# replace all faulty names to None

tweet_clean['name'].replace('the', 'None', inplace=True)
tweet_clean['name'].replace("light",'None', inplace=True)
tweet_clean['name'].replace("life",'None', inplace=True)
tweet_clean['name'].replace("an",'None', inplace=True)
tweet_clean['name'].replace("a",'None', inplace=True)
tweet_clean['name'].replace("his","None", inplace=True)
tweet_clean['name'].replace("by",'None', inplace=True)
tweet_clean['name'].replace("actually",'None', inplace=True)
tweet_clean['name'].replace("just",'None', inplace=True)
tweet_clean['name'].replace("getting",'None', inplace=True) 
tweet_clean['name'].replace("infuriating",'None', inplace=True) 
tweet_clean['name'].replace("old",'None', inplace=True) 
tweet_clean['name'].replace("all",'None', inplace=True) 
tweet_clean['name'].replace("this",'None', inplace=True) 
tweet_clean['name'].replace("very",'None', inplace=True) 
tweet_clean['name'].replace("not",'None', inplace=True)
tweet_clean['name'].replace("one",'None', inplace=True)
tweet_clean['name'].replace("my",'None', inplace=True)
tweet_clean['name'].replace("O","None", inplace=True)
tweet_clean['name'].replace("quite","None", inplace=True)
tweet_clean['name'].replace("such","None", inplace=True)

#### Test

In [111]:
tweet_clean.name.value_counts()

None        804
Charlie      11
Lucy         11
Oliver       10
Cooper       10
Penny         9
Tucker        9
Sadie         8
Lola          8
Winston       8
Daisy         7
Toby          7
Oscar         6
Jax           6
Bella         6
Bo            6
Bailey        6
Koda          6
Stanley       6
Buddy         5
Louis         5
Bentley       5
Dave          5
Leo           5
Milo          5
Chester       5
Rusty         5
Scout         5
Bear          4
Clarence      4
           ... 
Monty         1
Shawwn        1
Filup         1
Mason         1
Chubbs        1
Shaggy        1
Aiden         1
Trip          1
Blanket       1
Rumpole       1
Pablo         1
Peanut        1
Freddery      1
Bowie         1
Ralph         1
Amélie        1
Jeffri        1
Marlee        1
Dixie         1
Astrid        1
Benny         1
Tito          1
Zoe           1
Miguel        1
Pumpkin       1
Vince         1
Lilli         1
Julio         1
Major         1
Simba         1
Name: name, Length: 936,

#### Define
- Change retweeted_status_timestamp and timestamp to datetime objects and convert in_reply_to_status_id, in_reply_to_user_id, retweeted_status_id, retweeted_status_user_id to str objects

#### Clean

In [112]:
# correcting the tweets_clean datatypes

tweet_clean.timestamp = pd.to_datetime(tweet_clean.timestamp)
tweet_clean.retweeted_status_timestamp = pd.to_datetime(tweet_clean.retweeted_status_timestamp)
tweet_clean.in_reply_to_status_id = tweet_clean.in_reply_to_status_id.astype(str)
tweet_clean.in_reply_to_user_id = tweet_clean.in_reply_to_user_id.astype(str)
tweet_clean.retweeted_status_id = tweet_clean.retweeted_status_id.astype(str)
tweet_clean.retweeted_status_user_id = tweet_clean.retweeted_status_user_id.astype(str)

#### Test

In [113]:
tweet_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2200 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2200 non-null int64
in_reply_to_status_id         2200 non-null object
in_reply_to_user_id           2200 non-null object
timestamp                     2200 non-null datetime64[ns]
source                        2200 non-null object
text                          2200 non-null object
retweeted_status_id           2200 non-null object
retweeted_status_user_id      2200 non-null object
retweeted_status_timestamp    25 non-null datetime64[ns]
expanded_urls                 2141 non-null object
rating_numerator              2200 non-null int64
rating_denominator            2200 non-null int64
name                          2200 non-null object
doggo                         2200 non-null object
floofer                       2200 non-null object
pupper                        2200 non-null object
puppo                         2200 non-null object
dtypes: datetim

#### Define
- Remove tweets which has 0 denominator rating

#### Clean

In [114]:
drop_data = tweet_clean.query('rating_denominator <= 0')
tweet_clean.drop(drop_data.index, inplace=True)

#### Test

In [116]:
tweet_clean.rating_denominator.unique()

array([ 10,  15,  70,   7, 150,  11, 170,  20,  50,  90,  80,  40, 130,
       110,  16, 120,   2])

#### Define
- Remove tweets with missing tweet images.

#### Clean

In [117]:
tweet_clean = tweet_clean[tweet_clean.expanded_urls.notnull()]

#### Test

In [118]:
sum(tweet_clean.expanded_urls.isnull())

0

#### Define
- Make number of observations(rows) consistent across the tables

#### Clean

In [119]:
images_clean.shape, tweet_clean.shape, likes_clean.shape

((2075, 12), (2141, 17), (2341, 3))

In [120]:
# remove the rows whose data is deleted from WeRateDogs account
likes_tweet_ids = likes_clean.tweet_id
results = tweet_clean.query('tweet_id not in @likes_tweet_ids')
tweet_clean.drop(results.index, inplace=True)

In [121]:
images_clean.shape, tweet_clean.shape, likes_clean.shape

((2075, 12), (2137, 17), (2341, 3))

In [122]:
# retrive the tweet_ids that are present in tweets and images data
images_tweet_ids = images_clean.tweet_id
tweet_ids = tweet_clean.tweet_id
common_ids = set(tweet_ids) & set(images_tweet_ids)

In [123]:
# drop the rows whose image info is not provided
results = tweet_clean.query('tweet_id not in @common_ids')
tweet_clean.drop(results.index, inplace=True)
tweet_clean.shape

(2006, 17)

In [124]:
# drop the rows whose tweets data is not present
results = images_clean.query('tweet_id not in @common_ids')
images_clean.drop(results.index, inplace=True)
images_clean.shape

(2006, 12)

In [125]:
# make likes_df consitent to others
results = likes_clean.query('tweet_id not in @common_ids')
likes_clean.drop(results.index, inplace=True)
likes_clean.shape

(2006, 3)

#### Test

In [126]:
images_clean.shape, tweet_clean.shape, likes_clean.shape

((2006, 12), (2006, 17), (2006, 3))

#### Define
- Change None to NaN in columns 'name', 'doggo', 'floofer', 'pupper','puppo'

#### Clean

In [127]:
tweet_clean.name.replace('None', 'NaN', inplace=True)
tweet_clean.doggo.replace('None', 'NaN', inplace=True)
tweet_clean.floofer.replace('None', 'NaN', inplace=True)
tweet_clean.pupper.replace('None', 'NaN', inplace=True)
tweet_clean.puppo.replace('None', 'NaN', inplace=True)

#### Test

In [128]:
tweet_clean.name.unique()

array(['Phineas', 'Tilly', 'Archie', 'Darla', 'Franklin', 'NaN', 'Jax',
       'Zoey', 'Cassie', 'Koda', 'Bruno', 'Ted', 'Stuart', 'Oliver', 'Jim',
       'Zeke', 'Ralphus', 'Gerald', 'Jeffrey', 'Canela', 'Maya', 'Mingus',
       'Derek', 'Roscoe', 'Waffles', 'Jimbo', 'Maisey', 'Earl', 'Lola',
       'Kevin', 'Yogi', 'Noah', 'Bella', 'Grizzwald', 'Rusty', 'Gus',
       'Stanley', 'Alfy', 'Koko', 'Rey', 'Gary', 'Elliot', 'Louis',
       'Jesse', 'Romeo', 'Bailey', 'Duddles', 'Jack', 'Steven', 'Beau',
       'Snoopy', 'Shadow', 'Emmy', 'Aja', 'Penny', 'Dante', 'Nelly',
       'Ginger', 'Benedict', 'Venti', 'Goose', 'Nugget', 'Cash', 'Jed',
       'Sebastian', 'Sierra', 'Monkey', 'Harry', 'Kody', 'Lassie', 'Rover',
       'Napolean', 'Boomer', 'Cody', 'Rumble', 'Clifford', 'Dewey',
       'Scout', 'Gizmo', 'Walter', 'Cooper', 'Harold', 'Shikha', 'Lili',
       'Jamesy', 'Coco', 'Sammy', 'Meatball', 'Paisley', 'Albus',
       'Neptune', 'Belle', 'Quinn', 'Zooey', 'Dave', 'Jersey', 'Hobbes'

#### Define
- Change the rating_numerator and rating_denominator columns to float and assign the correct float values

#### Clean

In [129]:
# check if there are float values in text column
tweet_clean[tweet_clean.text.str.contains(r'(\d+(\.\d+))\/(\d+)')]['text'].values

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:2: UserWarning: This pattern has match groups. To actually get the groups, use str.extract.
  


array([ 'This is Bella. She hopes her smile made you smile. If not, she is also offering you her favorite monkey. 13.5/10 https://t.co/qjrljjt948',
       "This is Logan, the Chow who lived. He solemnly swears he's up to lots of good. H*ckin magical af 9.75/10 https://t.co/yBO5wuqaPS",
       "This is Sophie. She's a Jubilant Bush Pupper. Super h*ckin rare. Appears at random just to smile at the locals. 11.27/10 would smile back https://t.co/QFaUiIHxHq",
       'Here we have uncovered an entire battalion of holiday puppers. Average of 11.26/10 https://t.co/eNm2S6p9BD'], dtype=object)

In [130]:
# extract ratings from text column
ratings = tweet_clean['text'].str.extract(('(\d+(\.\d+)|(\d+))\/(\d+)'), expand=False)

In [131]:
# name ratings column
ratings.columns = ['num_float', 'float_decimal', 'num_interger', 'den_interger']

In [132]:
# check if correct float values are extracted from text column
ratings[ratings.num_float!=ratings.num_interger]

num_float float_decimal num_interger den_interger
45        13.5            .5          NaN           10
695       9.75           .75          NaN           10
763      11.27           .27          NaN           10
1712     11.26           .26          NaN           10

In [133]:
# copy 
tweet_clean['rating_numerator']=ratings.num_float.astype(float)
tweet_clean['rating_denominator']=ratings.den_interger.astype(float)

#### Test

In [134]:
tweet_clean.rating_numerator.unique()

array([  1.30000000e+01,   1.20000000e+01,   1.40000000e+01,
         1.35000000e+01,   1.10000000e+01,   1.00000000e+01,
         6.00000000e+00,   1.50000000e+01,   0.00000000e+00,
         8.40000000e+01,   2.40000000e+01,   9.75000000e+00,
         5.00000000e+00,   1.12700000e+01,   3.00000000e+00,
         7.00000000e+00,   8.00000000e+00,   9.00000000e+00,
         4.00000000e+00,   1.65000000e+02,   1.77600000e+03,
         2.04000000e+02,   5.00000000e+01,   9.90000000e+01,
         8.00000000e+01,   4.50000000e+01,   6.00000000e+01,
         4.40000000e+01,   1.43000000e+02,   1.21000000e+02,
         1.12600000e+01,   2.00000000e+00,   1.44000000e+02,
         8.80000000e+01,   1.00000000e+00,   4.20000000e+02])

In [135]:
tweet_clean.rating_denominator.unique()

array([  10.,   70.,    7.,  150.,   11.,  170.,   20.,   50.,   90.,
         80.,   40.,  130.,  110.,  120.,    2.])

In [136]:
tweet_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2006 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2006 non-null int64
in_reply_to_status_id         2006 non-null object
in_reply_to_user_id           2006 non-null object
timestamp                     2006 non-null datetime64[ns]
source                        2006 non-null object
text                          2006 non-null object
retweeted_status_id           2006 non-null object
retweeted_status_user_id      2006 non-null object
retweeted_status_timestamp    13 non-null datetime64[ns]
expanded_urls                 2006 non-null object
rating_numerator              2006 non-null float64
rating_denominator            2006 non-null float64
name                          2006 non-null object
doggo                         2006 non-null object
floofer                       2006 non-null object
pupper                        2006 non-null object
puppo                         2006 non-null object
dtypes: dat

#### Define
- Melt stages (doggo, floofer, pupper, puppo) into single column

#### Clean

In [137]:
tweet_clean = pd.melt(tweet_clean, id_vars=['tweet_id', 'in_reply_to_status_id', 'in_reply_to_user_id', 'timestamp', 'source', 'text',
                              'retweeted_status_id', 'retweeted_status_user_id', 'retweeted_status_timestamp',
                              'expanded_urls', 'rating_numerator', 'rating_denominator', 'name'],
        var_name='stage', value_vars=['doggo', 'floofer', 'pupper', 'puppo'], value_name = 'dog_stage'
       )


In [138]:
# drop the stage column after melting
tweet_clean.drop('stage', axis=1, inplace=True)

In [139]:
# drop duplicates caused due to melting
tweet_clean = tweet_clean.sort_values('dog_stage').drop_duplicates('tweet_id', keep = 'last')

#### Test

In [140]:
tweet_clean.dog_stage.value_counts()

NaN        1697
pupper      215
doggo        63
puppo        23
floofer       8
Name: dog_stage, dtype: int64

In [141]:
tweet_clean.shape

(2006, 14)

#### Define
- Move data from 3 dataframes to one master

#### Clean

In [142]:
# All three should have same number of rows as they are related to one tweet_id
images_clean.shape, tweet_clean.shape, likes_clean.shape

((2006, 12), (2006, 14), (2006, 3))

In [143]:
# merge tweet_clean and images_clean using tweet_id
master_df = tweet_clean.merge(images_clean, on='tweet_id')

In [144]:
master_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2006 entries, 0 to 2005
Data columns (total 25 columns):
tweet_id                      2006 non-null int64
in_reply_to_status_id         2006 non-null object
in_reply_to_user_id           2006 non-null object
timestamp                     2006 non-null datetime64[ns]
source                        2006 non-null object
text                          2006 non-null object
retweeted_status_id           2006 non-null object
retweeted_status_user_id      2006 non-null object
retweeted_status_timestamp    13 non-null datetime64[ns]
expanded_urls                 2006 non-null object
rating_numerator              2006 non-null float64
rating_denominator            2006 non-null float64
name                          2006 non-null object
dog_stage                     2006 non-null object
jpg_url                       2006 non-null object
img_num                       2006 non-null int64
p1                            2006 non-null object
p1_conf     

In [145]:
# merge master_df and likes_clean on tweet_id
master_df = master_df.merge(likes_clean, on='tweet_id')

#### Test

In [146]:
master_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2006 entries, 0 to 2005
Data columns (total 27 columns):
tweet_id                      2006 non-null int64
in_reply_to_status_id         2006 non-null object
in_reply_to_user_id           2006 non-null object
timestamp                     2006 non-null datetime64[ns]
source                        2006 non-null object
text                          2006 non-null object
retweeted_status_id           2006 non-null object
retweeted_status_user_id      2006 non-null object
retweeted_status_timestamp    13 non-null datetime64[ns]
expanded_urls                 2006 non-null object
rating_numerator              2006 non-null float64
rating_denominator            2006 non-null float64
name                          2006 non-null object
dog_stage                     2006 non-null object
jpg_url                       2006 non-null object
img_num                       2006 non-null int64
p1                            2006 non-null object
p1_conf     

In [147]:
master_df.head()

tweet_id in_reply_to_status_id in_reply_to_user_id  \
0  667435689202614272                   nan                 nan   
1  667437278097252352                   nan                 nan   
2  667443425659232256                   nan                 nan   
3  667453023279554560                   nan                 nan   
4  667455448082227200                   nan                 nan   

            timestamp                                             source  \
0 2015-11-19 20:14:03  <a href="http://twitter.com/download/iphone" r...   
1 2015-11-19 20:20:22  <a href="http://twitter.com/download/iphone" r...   
2 2015-11-19 20:44:47  <a href="http://twitter.com/download/iphone" r...   
3 2015-11-19 21:22:56  <a href="http://twitter.com" rel="nofollow">Tw...   
4 2015-11-19 21:32:34  <a href="http://twitter.com" rel="nofollow">Tw...   

                                                text retweeted_status_id  \
0            Ermergerd 12/10 https://t.co/PQni2sjPsm                 nan   
1  Never seen this breed before. Very pointy pup....                 nan   
2  Exotic dog here. Long neck. Weird paws. Obsess...                 nan   
3  Meet Cupcake. I would do unspeakable things fo...                 nan   
4  This is Reese and Twips. Reese protects Twips....                 nan   

  retweeted_status_user_id retweeted_status_timestamp  \
0                      nan                        NaT   
1                      nan                        NaT   
2                      nan                        NaT   
3                      nan                        NaT   
4                      nan                        NaT   

                                       expanded_urls      ...        p1_conf  \
0  https://twitter.com/dog_rates/status/667435689...      ...       0.999091   
1  https://twitter.com/dog_rates/status/667437278...      ...       0.989154   
2  https://twitter.com/dog_rates/status/667443425...      ...       0.980815   
3  https://twitter.com/dog_rates/status/667453023...      ...       0.825670   
4  https://twitter.com/dog_rates/status/667455448...      ...       0.676376   

   p1_dog                  p2   p2_conf p2_dog                         p3  \
0    True  miniature_pinscher  0.000450   True    black-and-tan_coonhound   
1   False          bath_towel  0.006300  False                     badger   
2   False               drake  0.006918  False                        hen   
3    True      French_bulldog  0.056639   True  Staffordshire_bullterrier   
4    True       Irish_terrier  0.054933   True          Yorkshire_terrier   

    p3_conf  p3_dog  favorite_count retweet_count  
0  0.000157    True             308            85  
1  0.000966   False             453           245  
2  0.005255   False             800           594  
3  0.054018    True             319            91  
4  0.040576    True             194            60  

[5 rows x 27 columns]

In [148]:
master_df.shape

(2006, 27)

### Store the cleaned dataset 

In [149]:
# write master_df to .csv file
master_df.to_csv('twitter_archive_master.csv', index=False)